nemo/collections/common/tokenizers/canary_tokenizer.py

```diff
@@ -69,22 +71,34 @@ def pad_id(self) -> int:
     def text_to_ids(self, text, lang_id) -> list[int]:
         if lang_id == CANARY_SPECIAL_TOKENIZER:
             return self._tokenize_special_prompt(text)
+        lang_id = _map_canary1_to_canary2_lang(lang_id, self.langs)
         if text.endswith(CANARY_EOS):
             return super().text_to_ids(text[: -len(CANARY_EOS)], lang_id) + [self.eos_id]
         return super().text_to_ids(text, lang_id)
 
     def _tokenize_special_prompt(self, text: str) -> list[int]:
         """
-        Tokenize the input special prompt of the following schema:
-
-        <|startoftranscript|><|source_lang|><|taskname|><|target_lang|><|pnc|>
+        Tokenize the input special prompt of Canary family of models.
 
         Required because otherwise self.text_to_ids() returns a different result than what Canary had been trained with.
         """
         ans = []
-        assert text.count('>') == 5, f"Expected exactly 5 special tokens in Canary's prompt, got: {text}."
-        assert text.startswith(CANARY_BOS), text
-        for _ in range(5):
+
+        if text.startswith(CANARY2_BOCTX):
```

Comment:
Wait, can you explain this? I don't remember this in tokenizer discussions why this needs to be the case.

Reply:
this is for decoding with context. the way canary currently does long-form transcripttion is to divide audio into non-overlapping chunks and run inference as independent examples. this is to be able to feed previous chunks inference into current chunks inference (and potentially explore better recognition of names if they were given in context)

Reply:
Oh that makes way more sense